flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/ExecutionAttemptMappingProvider.java

```diff
@@ -56,14 +56,16 @@ protected boolean removeEldestEntry(
     }
 
     public Optional<ExecutionVertex> getVertex(ExecutionAttemptID id) {
-        if (!cachedTasksById.containsKey(id)) {
-            cachedTasksById.putAll(getCurrentAttemptMappings());
+        synchronized (cachedTasksById) {
```

Comment:
I think it is possible to do locking in a more optimal way(ConcurrentHashMap + something) but it will complicate the solution and right now I am not sure that it makes sense to do so.

Reply:
I am thinking if the entire caching here is not overly complicated. As far as I understand it, we're rebuilding the entire `cachedTasksById` on each miss. Wouldn't it make more sense to do so explicitly? What do you think about code as follows?

```
    public Optional<ExecutionVertex> getVertex(ExecutionAttemptID id) {
        ExecutionVertex vertex = cachedTasksById.get(id);
        if (vertex != null) {
            return Optional.of(vertex);
        } else {
            return updateAndGet(id);
        }
    }

    private synchronized Optional<ExecutionVertex> updateAndGet(ExecutionAttemptID id) {
        ExecutionVertex vertex;
        // might've been updated by another thread
        vertex = cachedTasksById.get(id);
        if (vertex != null) {
            return Optional.of(vertex);
        }
        cachedTasksById = getCurrentAttemptMappings();
        return Optional.ofNullable(cachedTasksById.get(id));
    }
```

Reply:
Unfortunately, it is not a fully correct solution since  `cachedTasksById.containsKey(id)` and `cachedTasksById.get(id) == null` can be both true at the same time so you can not replace `containsKey` by null check. Secondly, originally we have `LinkedHashMap` and remove the oldest entry if the size is exceeded while for concurrent implementation it will be more tricky.
In fact, we can resolve all problems but as I said before it will be more complicated (ConcurrentHashMap + ConcurrentLinkedQueue + NullObject). 
Right now I don't really understand how critical this code is and how often it is executed. But maybe I will try my idea and we will see how complicated it looks like

Reply:
Now I see your point, but the current implementation is able to return `null` from the cache if it missed the cache once while the proposed implementation will recreate cache every time on the miss. I mean these lines:
```
if (!cachedTasksById.containsKey(id)) {
                    // the task probably gone after a restart
                    cachedTasksById.put(id, null);
}
```
I actually don't fully understand the true purpose of this. I mean do we really need to cache one last id which missed the cache? How is it likely that we request it again? Do you understand the purpose of that?

Reply:
> Now I see your point, but the current implementation is able to return null from the cache if it missed the cache once while the proposed implementation will recreate cache every time on the miss. I mean these lines:

Yes, that's something I missed, that we treat `null` in a special way. I need to give it a second thought.

> Oh, or do you mean that we can drop all mapping when we have even one miss?

I might be wrong, but I think that's the purpose of overwriting the method of `LinkedHashMap` in the ctor.

Reply:
I looked into the code and I think it is rather unlikely we would request again an id that caused the miss. As far as I can tell it is possible if we had multiple aborted checkpoints on a single task manager which is rather unlikely. However I am not 100% sure if there are no other hidden reasons for it...

If we want to maintain the special treatment of `null` I think we can do:
```
    public Optional<ExecutionVertex> getVertex(ExecutionAttemptID id) {
        if (cachedTasksById.containsKey(id)) {
            return Optional.ofNullable(cachedTasksById.get(id));
        }

        return updateAndGet(id);
    }

    private Optional<ExecutionVertex> updateAndGet(ExecutionAttemptID id) {
        synchronized (tasks) {
            if (cachedTasksById.containsKey(id)) {
                return Optional.ofNullable(cachedTasksById.get(id));
            }

            cachedTasksById = getCurrentAttemptMappings();
            if (!cachedTasksById.containsKey(id)) {
                cachedTasksById.put(id, null);
                return Optional.empty();
            }
            return Optional.ofNullable(cachedTasksById.get(id));
        }
    }
```

Reply:
I changed the implementation with `cachedTasksById.put(id, null);`